cmd/jag/commands/watch.go

```diff
@@ -78,9 +68,142 @@ func WatchCmd() *cobra.Command {
 	return cmd
 }
 
-func onWatchChanges(ctx context.Context, watcher *fsnotify.Watcher, device *Device, sdk *SDK, entrypoint string) (<-chan struct{}, func()) {
+type watcher struct {
+	sync.Mutex
+	watcher *fsnotify.Watcher
+
+	paths map[string]struct{}
+}
+
+func newWatcher() (*watcher, error) {
+	w, err := fsnotify.NewWatcher()
+	if err != nil {
+		return nil, err
+	}
+	return &watcher{
+		watcher: w,
+		paths:   map[string]struct{}{},
+	}, nil
+}
+
+func (w *watcher) Close() error {
+	return w.Close()
+}
+
+func (w *watcher) Events() chan fsnotify.Event {
+	return w.watcher.Events
+}
+
+func (w *watcher) Errors() chan error {
+	return w.watcher.Errors
+}
+
+func (w *watcher) Watch(paths ...string) (err error) {
+	for i, p := range paths {
+		if paths[i], err = filepath.EvalSymlinks(p); err != nil {
+			return err
+		}
+	}
+
+	paths = findParents(paths...)
+	fmt.Println("Watching paths: ", paths)
+	candidates := map[string]struct{}{}
+	for _, p := range paths {
+		if _, ok := w.paths[p]; !ok {
+			w.Mutex.Lock()
+			fmt.Println("watch path", p)
+			w.watcher.Add(p)
+			w.paths[p] = struct{}{}
+			w.Mutex.Unlock()
+		}
+		candidates[p] = struct{}{}
+	}
+
+	for p := range w.paths {
+		if _, ok := candidates[p]; !ok {
+			w.Mutex.Lock()
+			fmt.Println("remove path", p)
+			w.watcher.Remove(p)
+			delete(w.paths, p)
+			w.Mutex.Unlock()
+		}
+	}
+	return nil
+}
+
+func findParents(paths ...string) []string {
+	var res []string
+	for _, p := range paths {
+		if len(res) == 0 {
+			res = append(res, p)
+			continue
+		}
+
+		matched := false
+		for i, r := range res {
+			fmt.Printf("Is %s sub of %s: %v\n", r, p, isSub(r, p))
+			if isSub(r, p) {
+				matched = true
+				break
+			}
+			fmt.Printf("Is %s sub of %s: %v\n", p, r, isSub(p, r))
+			if isSub(p, r) {
+				matched = true
+				res[i] = p
+				break
+			}
+		}
+		if !matched {
+			res = append(res, p)
+		}
+	}
+	return res
+}
+
+func isSub(parent, sub string) bool {
+	if parent == sub {
+		return true
+	}
+	pattern := parent + string(filepath.Separator) + "*"
+	if ok, err := filepath.Match(pattern, sub); err == nil && ok {
+		return true
+	}
+	return false
+}
+
+func parseDependeniesToDirs(b []byte) []string {
+	m := map[string]struct{}{}
+	scanner := bufio.NewScanner(bytes.NewReader(b))
+	for scanner.Scan() {
+		p := strings.TrimSuffix(strings.TrimSpace(scanner.Text()), ":")
+		m[filepath.Dir(p)] = struct{}{}
+	}
+	var res []string
+	for r := range m {
+		res = append(res, r)
+	}
+	return res
+}
+
+func onWatchChanges(ctx context.Context, watcher *watcher, device *Device, sdk *SDK, entrypoint string) (<-chan struct{}, func()) {
 	doneCh := make(chan struct{})
 
+	updateWatcher := func(runCtx context.Context) {
+		cmd := sdk.Toitc(ctx, "--dependency-file", "-", "--dependency-format", "plain", "--analyze", entrypoint)
```

Comment:
That feels dangerous to me.
I would need to check, but I would expect the `--analyze` to still emit warnings on stdout.

Reply:
Verified it: I don't think there is an easy way to avoid warnings in the output at the moment.
The safest would be to use a temporary file for the dependencies.